File: krishideep_app/lib/services/crop_service.dart
import 'dart:math' as math;
import 'dart:io';
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;
import '../models/crop.dart';
import '../models/crop_advice.dart';
import '../models/crop_recommendation.dart';
import '../models/disease_detection.dart';
import '../models/soil_data.dart';
import 'soil_data_service.dart';

class CropService {
  // Real API-based crop data only

  // Get crop recommendation using real API
  Future<CropRecommendation> getCropRecommendationFromAPI(
    CropAdviceRequest request,
  ) async {
    try {
      // Convert request to SoilData
      final soilData = request.toSoilData();

      // Validate soil data
      if (!SoilDataService.validateSoilData(soilData)) {
        throw Exception('Invalid soil data provided');
      }

      // Get crop prediction from API
      final prediction = await SoilDataService.getCropPrediction(soilData);

      if (prediction != null) {
        // Create crop object from prediction
        final recommendedCrop = Crop(
          id: 'api_${DateTime.now().millisecondsSinceEpoch}',
          name: prediction.cropName,
          nameHindi: _getHindiName(prediction.cropName),
          description: prediction.description ??
              'AI-recommended crop based on soil analysis',
          expectedYield: _estimateYield(prediction.cropName, soilData),
          profitMargin: _estimateProfitMargin(prediction.cropName),
          sustainabilityScore: _calculateSustainabilityScore(soilData),
          suitableConditions:
              _getSuitableConditions(prediction.cropName, soilData),
        );

        // Get soil recommendations
        final soilRecommendations =
            SoilDataService.getSoilRecommendations(soilData);

        return CropRecommendation(
          recommendedCrops: [recommendedCrop],
          recommendation:
              _generateAPIBasedRecommendation(prediction, soilData, request),
          confidenceLevel: _getConfidenceLevel(prediction.confidence),
          timestamp: DateTime.now(),
          analysisData: {
            'soilPH': soilData.ph,
            'soilMoisture': soilData.soilMoistureAvg,
            'temperature': soilData.temperature,
            'humidity': soilData.humidity,
            'rainfall': soilData.rainfall,
            'nitrogen': soilData.n,
            'phosphorus': soilData.p,
            'potassium': soilData.k,
            'location': request.location,
            'confidence': prediction.confidence,
            'soilRecommendations': soilRecommendations,
          },
        );
      } else {
        // Fallback to local recommendation if API fails
        return await getCropRecommendation(request);
      }
    } catch (e) {
      print('API-based recommendation failed: $e');
      // Fallback to local recommendation
      return await getCropRecommendation(request);
    }
  }

  // Real API-based crop recommendation only
  Future<CropRecommendation> getCropRecommendation(
    CropAdviceRequest request,
  ) async {
    // Use only real API-based recommendations
    return await getCropRecommendationFromAPI(request);
  }

  // Real API-based methods only

  // Real disease detection using ML API
  Future<DiseaseDetectionResult> detectDisease(File imageFile) async {
    print('🔍 DEBUG: CropService.detectDisease() called');
    print('🔍 DEBUG: Image file path: ${imageFile.path}');
    print('🔍 DEBUG: Image file exists: ${await imageFile.exists()}');
    print('🔍 DEBUG: Image file size: ${await imageFile.length()} bytes');

    try {
      print('🔍 DEBUG: Making API call to disease detection service...');

      // Create multipart request with correct endpoint
      var request = http.MultipartRequest(
        'POST',
        Uri.parse('https://backend-krisi-ml-crophealth.onrender.com/predict'),
      );

      // Add image file with 'file' as key (as specified by user)
      request.files.add(await http.MultipartFile.fromPath(
        'file',
        imageFile.path,
      ));

      print('🔍 DEBUG: Sending request to API with form data...');
      print(
          '🔍 DEBUG: API endpoint: https://backend-krisi-ml-crophealth.onrender.com/predict');
      print('🔍 DEBUG: Form field key: file');
      print('🔍 DEBUG: File path: ${imageFile.path}');
      print('🔍 DEBUG: Request method: POST');
      print('🔍 DEBUG: Content-Type: multipart/form-data');
      print('🔍 DEBUG: File size: ${await imageFile.length()} bytes');
      print('🔍 DEBUG: File exists: ${await imageFile.exists()}');

      var response = await request.send().timeout(
        const Duration(seconds: 60), // Increased timeout for ML processing
        onTimeout: () {
          throw TimeoutException('API request timed out after 60 seconds',
              const Duration(seconds: 60));
        },
      );

      print('🔍 DEBUG: API response status: ${response.statusCode}');
      print('🔍 DEBUG: API response headers: ${response.headers}');

      if (response.statusCode == 200) {
        print('🔍 DEBUG: API call successful, parsing response...');
        var responseBody = await response.stream.bytesToString();
        print('🔍 DEBUG: Raw API response: $responseBody');

        var jsonResponse = json.decode(responseBody);
        print('🔍 DEBUG: Parsed JSON response: $jsonResponse');

        // Parse the API response with proper error handling
        // Based on the working HTML code, the API returns 'prediction' field
        String prediction = jsonResponse['prediction'] ??
            jsonResponse['disease_name'] ??
            jsonResponse['diseaseName'] ??
            'Unknown Disease';

        return DiseaseDetectionResult(
          diseaseName: prediction,
          diseaseNameHindi: _getHindiName(prediction),
          description:
              jsonResponse['description'] ?? 'AI-detected disease: $prediction',
          confidence: (jsonResponse['confidence'] ?? 85.0)
              .toDouble(), // Default confidence if not provided
          severity: jsonResponse['severity'] ??
              _getSeverityFromPrediction(prediction),
          treatments: List<String>.from(
              jsonResponse['treatments'] ?? _getDefaultTreatments(prediction)),
          prevention: List<String>.from(
              jsonResponse['prevention'] ?? _getDefaultPrevention(prediction)),
        );
      } else {
        print('🔍 DEBUG: API call failed with status: ${response.statusCode}');
        var errorBody = await response.stream.bytesToString();
        print('🔍 DEBUG: API error response: $errorBody');
        throw Exception(
            'API call failed with status: ${response.statusCode}. Error: $errorBody');
      }
    } catch (e) {
      print('🔍 DEBUG: API call failed with error: $e');
      print('🔍 DEBUG: Error type: ${e.runtimeType}');
      print('🔍 DEBUG: API Error Details: $e');

      // Create a result that shows the API error instead of mock data
      return DiseaseDetectionResult(
        diseaseName: 'API Error',
        diseaseNameHindi: 'API त्रुटि',
        description: 'API call failed: $e',
        confidence: 0.0,
        severity: 'Error',
        treatments: [
          'Check internet connection',
          'Try again later',
          'Contact support'
        ],
        prevention: ['Ensure stable internet', 'Check API status'],
      );
    }
  }

  // No mock data - only real API results

  // Real API-based crop data only

  // Helper methods for API-based recommendations
  String _getHindiName(String englishName) {
    final hindiNames = {
      'Rice': 'चावल',
      'Wheat': 'गेहूं',
      'Maize': 'मक्का',
      'Sugarcane': 'गन्ना',
      'Cotton': 'कपास',
      'Potato': 'आलू',
      'Tomato': 'टमाटर',
      'Onion': 'प्याज',
      'Chili': 'मिर्च',
      'Brinjal': 'बैंगन',
      'Cabbage': 'पत्ता गोभी',
      'Cauliflower': 'फूल गोभी',
      'Carrot': 'गाजर',
      'Radish': 'मूली',
      'Spinach': 'पालक',
      'Coriander': 'धनिया',
      'Mint': 'पुदीना',
      'Basil': 'तुलसी',
      'Ginger': 'अदरक',
      'Turmeric': 'हल्दी',
    };
    return hindiNames[englishName] ?? englishName;
  }

  double _estimateYield(String cropName, SoilData soilData) {
    final baseYields = {
      'Rice': 2500.0,
      'Wheat': 1800.0,
      'Maize': 3200.0,
      'Sugarcane': 45000.0,
      'Cotton': 800.0,
      'Potato': 15000.0,
      'Tomato': 20000.0,
      'Onion': 12000.0,
    };

    double baseYield = baseYields[cropName] ?? 1000.0;

    // Adjust yield based on soil conditions
    double multiplier = 1.0;

    // pH adjustment
    if (soilData.ph >= 6.0 && soilData.ph <= 7.5) {
      multiplier += 0.2;
    } else if (soilData.ph < 5.0 || soilData.ph > 8.5) {
      multiplier -= 0.3;
    }

    // Nutrient adjustment
    if (soilData.n >= 30 && soilData.n <= 100) multiplier += 0.1;
    if (soilData.p >= 20 && soilData.p <= 80) multiplier += 0.1;
    if (soilData.k >= 15 && soilData.k <= 60) multiplier += 0.1;

    // Temperature adjustment
    if (soilData.temperature >= 20 && soilData.temperature <= 30) {
      multiplier += 0.15;
    } else if (soilData.temperature < 10 || soilData.temperature > 40) {
      multiplier -= 0.2;
    }

    return (baseYield * multiplier).roundToDouble();
  }

  double _estimateProfitMargin(String cropName) {
    final profitMargins = {
      'Rice': 25.5,
      'Wheat': 30.2,
      'Maize': 35.8,
      'Sugarcane': 28.5,
      'Cotton': 32.1,
      'Potato': 40.0,
      'Tomato': 45.0,
      'Onion': 35.0,
    };
    return profitMargins[cropName] ?? 30.0;
  }

  double _calculateSustainabilityScore(SoilData soilData) {
    double score = 5.0; // Base score

    // pH sustainability
    if (soilData.ph >= 6.0 && soilData.ph <= 7.5) score += 1.0;

    // Nutrient balance
    if (soilData.n >= 30 && soilData.n <= 100) score += 0.5;
    if (soilData.p >= 20 && soilData.p <= 80) score += 0.5;
    if (soilData.k >= 15 && soilData.k <= 60) score += 0.5;

    // Environmental factors
    if (soilData.temperature >= 15 && soilData.temperature <= 35) score += 0.5;
    if (soilData.humidity >= 40 && soilData.humidity <= 80) score += 0.5;
    if (soilData.rainfall >= 50 && soilData.rainfall <= 200) score += 0.5;

    return math.min(score, 10.0);
  }

  List<String> _getSuitableConditions(String cropName, SoilData soilData) {
    List<String> conditions = [];

    // pH conditions
    if (soilData.ph >= 6.0 && soilData.ph <= 7.5) {
      conditions.add('Optimal pH (${soilData.ph.toStringAsFixed(1)})');
    } else {
      conditions
          .add('pH adjustment needed (${soilData.ph.toStringAsFixed(1)})');
    }

    // Temperature conditions
    if (soilData.temperature >= 20 && soilData.temperature <= 30) {
      conditions.add(
          'Ideal temperature (${soilData.temperature.toStringAsFixed(1)}°C)');
    } else {
      conditions.add(
          'Temperature monitoring needed (${soilData.temperature.toStringAsFixed(1)}°C)');
    }

    // Moisture conditions
    if (soilData.soilMoistureAvg >= 30 && soilData.soilMoistureAvg <= 70) {
      conditions.add(
          'Good soil moisture (${soilData.soilMoistureAvg.toStringAsFixed(1)}%)');
    } else {
      conditions.add(
          'Moisture management needed (${soilData.soilMoistureAvg.toStringAsFixed(1)}%)');
    }

    // Nutrient conditions
    if (soilData.n >= 30 && soilData.n <= 100) {
      conditions.add('Adequate nitrogen (${soilData.n.toStringAsFixed(1)})');
    } else {
      conditions
          .add('Nitrogen adjustment needed (${soilData.n.toStringAsFixed(1)})');
    }

    return conditions;
  }

  String _getConfidenceLevel(double confidence) {
    if (confidence >= 80) return 'High';
    if (confidence >= 60) return 'Medium';
    return 'Low';
  }

  String _generateAPIBasedRecommendation(
      CropPrediction prediction, SoilData soilData, CropAdviceRequest request) {
    return 'Based on comprehensive soil analysis using AI prediction model, '
        '${prediction.cropName} is recommended with ${prediction.confidence.toStringAsFixed(1)}% confidence. '
        'Your soil conditions (pH: ${soilData.ph.toStringAsFixed(1)}, '
        'Temperature: ${soilData.temperature.toStringAsFixed(1)}°C, '
        'Moisture: ${soilData.soilMoistureAvg.toStringAsFixed(1)}%) '
        'are suitable for this crop. '
        'Consider the soil recommendations for optimal yield.';
  }

  // Helper methods for disease detection based on prediction
  String _getSeverityFromPrediction(String prediction) {
    final severeDiseases = ['blight', 'wilt', 'rot', 'canker', 'rust'];
    final moderateDiseases = ['spot', 'mildew', 'mosaic', 'yellow'];

    String lowerPrediction = prediction.toLowerCase();

    for (String disease in severeDiseases) {
      if (lowerPrediction.contains(disease)) return 'Severe';
    }

    for (String disease in moderateDiseases) {
      if (lowerPrediction.contains(disease)) return 'Moderate';
    }

    if (lowerPrediction.contains('healthy') ||
        lowerPrediction.contains('normal')) {
      return 'None';
    }

    return 'Mild';
  }

  List<String> _getDefaultTreatments(String prediction) {
    String lowerPrediction = prediction.toLowerCase();

    if (lowerPrediction.contains('healthy') ||
        lowerPrediction.contains('normal')) {
      return [
        'Continue regular care',
        'Maintain proper nutrition',
        'Monitor regularly',
      ];
    }

    if (lowerPrediction.contains('blight')) {
      return [
        'Apply copper-based fungicide',
        'Remove infected plant parts',
        'Improve air circulation',
        'Avoid overhead watering',
      ];
    }

    if (lowerPrediction.contains('mildew')) {
      return [
        'Apply sulfur-based fungicide',
        'Use neem oil spray',
        'Improve air circulation',
        'Reduce humidity',
      ];
    }

    if (lowerPrediction.contains('spot')) {
      return [
        'Apply fungicide spray',
        'Remove affected leaves',
        'Improve drainage',
        'Avoid wetting leaves',
      ];
    }

    // Default treatments for unknown diseases
    return [
      'Consult with agricultural expert',
      'Apply appropriate fungicide',
      'Remove infected parts',
      'Improve plant care',
    ];
  }

  List<String> _getDefaultPrevention(String prediction) {
    String lowerPrediction = prediction.toLowerCase();

    if (lowerPrediction.contains('healthy') ||
        lowerPrediction.contains('normal')) {
      return [
        'Continue good practices',
        'Regular monitoring',
        'Proper nutrition',
        'Maintain hygiene',
      ];
    }

    return [
      'Use disease-resistant varieties',
      'Practice crop rotation',
      'Maintain proper spacing',
      'Avoid overhead watering',
      'Regular field inspection',
      'Good drainage',
      'Balanced fertilization',
    ];
  }
}

File: krisideepweb/src/services/CropService.ts
import { Crop, CropRecommendation, CropAdviceRequest, DiseaseDetectionResult } from '../models';

class CropService {
  // Comprehensive crop data for demonstration
  private dummyCrops: Crop[] = [
    {
      id: '1',
      name: 'Rice',
      nameHindi: 'चावल',
      description: 'Staple crop suitable for humid regions with good water availability. High yielding varieties available.',
      expectedYield: 2800, // kg per acre
      profitMargin: 28.5,
      sustainabilityScore: 8.2,
      suitableConditions: ['High moisture (60-80%)', 'pH 6.0-7.5', 'Fertile alluvial soil', 'Temperature 20-35°C'],
      imageUrl: '/images/rice.jpg',
    },
    {
      id: '2',
      name: 'Wheat',
      nameHindi: 'गेहूं',
      description: 'Winter crop suitable for cooler regions with moderate water needs. High protein content.',
      expectedYield: 2200,
      profitMargin: 32.8,
      sustainabilityScore: 8.5,
      suitableConditions: ['Moderate moisture (40-60%)', 'pH 6.5-7.5', 'Well-drained soil', 'Temperature 15-25°C'],
      imageUrl: '/images/wheat.jpg',
    },
    {
      id: '3',
      name: 'Maize',
      nameHindi: 'मक्का',
      description: 'Versatile crop suitable for various climatic conditions. High energy content.',
      expectedYield: 3800,
      profitMargin: 38.2,
      sustainabilityScore: 7.8,
      suitableConditions: ['Moderate moisture (50-70%)', 'pH 6.0-7.0', 'Fertile soil', 'Temperature 18-30°C'],
      imageUrl: '/images/maize.jpg',
    },
    {
      id: '4',
      name: 'Sugarcane',
      nameHindi: 'गन्ना',
      description: 'Cash crop requiring high water and warm climate. High sugar content.',
      expectedYield: 52000,
      profitMargin: 31.5,
      sustainabilityScore: 6.8,
      suitableConditions: ['High moisture (70-90%)', 'pH 6.5-7.5', 'Rich fertile soil', 'Temperature 25-35°C'],
      imageUrl: '/images/sugarcane.jpg',
    },
    {
      id: '5',
      name: 'Cotton',
      nameHindi: 'कपास',
      description: 'Fiber crop suitable for black cotton soil regions. High market value.',
      expectedYield: 950,
      profitMargin: 35.8,
      sustainabilityScore: 7.2,
      suitableConditions: ['Moderate moisture (40-60%)', 'pH 6.0-8.0', 'Black cotton soil', 'Temperature 20-30°C'],
      imageUrl: '/images/cotton.jpg',
    },
    {
      id: '6',
      name: 'Soybean',
      nameHindi: 'सोयाबीन',
      description: 'Protein-rich legume crop. Good for soil nitrogen fixation.',
      expectedYield: 1800,
      profitMargin: 29.5,
      sustainabilityScore: 8.8,
      suitableConditions: ['Moderate moisture (50-70%)', 'pH 6.0-7.0', 'Well-drained soil', 'Temperature 20-30°C'],
      imageUrl: '/images/soybean.jpg',
    },
    {
      id: '7',
      name: 'Groundnut',
      nameHindi: 'मूंगफली',
      description: 'Oilseed crop with high nutritional value. Drought tolerant.',
      expectedYield: 1200,
      profitMargin: 33.2,
      sustainabilityScore: 8.0,
      suitableConditions: ['Low to moderate moisture (30-50%)', 'pH 6.0-7.5', 'Sandy loam soil', 'Temperature 25-35°C'],
      imageUrl: '/images/groundnut.jpg',
    },
    {
      id: '8',
      name: 'Tomato',
      nameHindi: 'टमाटर',
      description: 'High-value vegetable crop. Short duration and high yield potential.',
      expectedYield: 25000,
      profitMargin: 45.8,
      sustainabilityScore: 7.5,
      suitableConditions: ['Moderate moisture (50-70%)', 'pH 6.0-7.0', 'Fertile soil', 'Temperature 20-30°C'],
      imageUrl: '/images/tomato.jpg',
    },
    {
      id: '9',
      name: 'Potato',
      nameHindi: 'आलू',
      description: 'Staple vegetable crop. High carbohydrate content and good storage.',
      expectedYield: 18000,
      profitMargin: 28.5,
      sustainabilityScore: 7.8,
      suitableConditions: ['Moderate moisture (60-80%)', 'pH 5.5-6.5', 'Sandy loam soil', 'Temperature 15-25°C'],
      imageUrl: '/images/potato.jpg',
    },
    {
      id: '10',
      name: 'Onion',
      nameHindi: 'प्याज',
      description: 'Essential vegetable crop. Good market demand and storage life.',
      expectedYield: 15000,
      profitMargin: 42.3,
      sustainabilityScore: 8.1,
      suitableConditions: ['Moderate moisture (50-70%)', 'pH 6.0-7.0', 'Well-drained soil', 'Temperature 15-25°C'],
      imageUrl: '/images/onion.jpg',
    },
  ];

  // Simulate AI-based crop recommendation
  async getCropRecommendation(request: CropAdviceRequest): Promise<CropRecommendation> {
    // Simulate API delay
    await new Promise(resolve => setTimeout(resolve, 2000));

    // Simple rule-based recommendation logic (replace with actual AI model)
    const recommendedCrops: Crop[] = [];

    for (const crop of this.dummyCrops) {
      const suitabilityScore = this.calculateSuitabilityScore(crop, request);
      if (suitabilityScore > 0.6) {
        // Modify crop data based on request parameters
        recommendedCrops.push({
          ...crop,
          expectedYield: crop.expectedYield * suitabilityScore,
          profitMargin: crop.profitMargin * suitabilityScore,
        });
      }
    }

    // Sort by expected yield
    recommendedCrops.sort((a, b) => b.expectedYield - a.expectedYield);

    // Take top 3 recommendations
    const topRecommendations = recommendedCrops.slice(0, 3);

    const recommendation = this.generateRecommendationText(request, topRecommendations);
    const confidenceLevel = topRecommendations.length > 0 ? 'High' : 'Low';

    return {
      recommendedCrops: topRecommendations,
      recommendation,
      confidenceLevel,
      timestamp: new Date(),
      analysisData: {
        soilPH: request.soilPH,
        soilMoisture: request.soilMoisture,
        farmSize: request.farmSize,
        location: request.location,
      },
    };
  }

  // Calculate suitability score based on soil conditions
  private calculateSuitabilityScore(crop: Crop, request: CropAdviceRequest): number {
    let score = 0.5; // Base score

    // pH suitability
    if (crop.name === 'Rice' && request.soilPH >= 6.0 && request.soilPH <= 7.5) {
      score += 0.2;
    } else if (crop.name === 'Wheat' && request.soilPH >= 6.5 && request.soilPH <= 7.5) {
      score += 0.2;
    } else if (crop.name === 'Maize' && request.soilPH >= 6.0 && request.soilPH <= 7.0) {
      score += 0.2;
    } else if (crop.name === 'Sugarcane' && request.soilPH >= 6.5 && request.soilPH <= 7.5) {
      score += 0.2;
    } else if (crop.name === 'Cotton' && request.soilPH >= 6.0 && request.soilPH <= 8.0) {
      score += 0.2;
    }

    // Moisture suitability
    if ((crop.name === 'Rice' || crop.name === 'Sugarcane') && request.soilMoisture > 60) {
      score += 0.2;
    } else if (
      (crop.name === 'Wheat' || crop.name === 'Maize' || crop.name === 'Cotton') &&
      request.soilMoisture >= 30 &&
      request.soilMoisture <= 70
    ) {
      score += 0.2;
    }

    // Farm size consideration
    if (request.farmSize > 5) {
      score += 0.1;
    }

    return Math.min(score, 1.0);
  }

  // Generate recommendation text
  private generateRecommendationText(request: CropAdviceRequest, crops: Crop[]): string {
    if (crops.length === 0) {
      return `Based on your soil conditions (pH: ${request.soilPH}, Moisture: ${request.soilMoisture}%), we recommend consulting with local agricultural experts for personalized advice.`;
    }

    const topCrop = crops[0].name;
    return `Based on your soil analysis (pH: ${request.soilPH}, Moisture: ${request.soilMoisture}%), ${topCrop} appears to be the most suitable crop for your ${request.farmSize}-acre farm. The recommended crops are well-suited for your soil conditions and climate.`;
  }

  // Simulate disease detection using ML model
  async detectDisease(imageFile: File): Promise<DiseaseDetectionResult> {
    // Simulate processing delay
    await new Promise(resolve => setTimeout(resolve, 3000));

    // Comprehensive disease detection results
    const possibleDiseases: DiseaseDetectionResult[] = [
      {
        diseaseName: 'Bacterial Leaf Blight',
        diseaseNameHindi: 'जीवाणु पत्ती झुलसा',
        description: 'A serious bacterial disease causing water-soaked lesions on leaves, leading to significant yield loss in rice crops.',
        confidence: 89.2,
        severity: 'High',
        treatments: [
          'Apply copper-based bactericides',
          'Remove and destroy infected plants',
          'Improve field drainage',
          'Use resistant varieties in next season',
          'Apply streptomycin spray (0.1%)',
        ],
        prevention: [
          'Use certified disease-free seeds',
          'Maintain proper field hygiene',
          'Avoid overhead irrigation',
          'Practice crop rotation',
          'Monitor weather conditions',
        ],
        imageUrl: '/images/bacterial-blight.jpg',
      },
      {
        diseaseName: 'Powdery Mildew',
        diseaseNameHindi: 'चूर्णी फफूंदी',
        description: 'Fungal disease causing white powdery coating on leaves, reducing photosynthesis and plant vigor.',
        confidence: 85.7,
        severity: 'Moderate',
        treatments: [
          'Apply sulfur-based fungicides',
          'Spray neem oil solution (2%)',
          'Use baking soda spray (1 tsp per liter)',
          'Remove heavily infected leaves',
          'Apply potassium bicarbonate',
        ],
        prevention: [
          'Ensure good air circulation',
          'Avoid overcrowding plants',
          'Water at soil level only',
          'Use resistant varieties',
          'Regular field monitoring',
        ],
        imageUrl: '/images/powdery-mildew.jpg',
      },
      {
        diseaseName: 'Rust Disease',
        diseaseNameHindi: 'रस्ट रोग',
        description: 'Fungal disease causing orange or brown pustules on leaves, stems, and grains, affecting crop quality.',
        confidence: 91.3,
        severity: 'High',
        treatments: [
          'Apply propiconazole fungicide',
          'Remove infected plant debris',
          'Improve field ventilation',
          'Use systemic fungicides',
          'Apply mancozeb spray',
        ],
        prevention: [
          'Plant resistant varieties',
          'Maintain proper spacing',
          'Avoid excessive nitrogen',
          'Practice crop rotation',
          'Monitor humidity levels',
        ],
        imageUrl: '/images/rust-disease.jpg',
      },
      {
        diseaseName: 'Anthracnose',
        diseaseNameHindi: 'एन्थ्रेक्नोज',
        description: 'Fungal disease causing dark, sunken lesions on fruits, leaves, and stems, leading to fruit rot.',
        confidence: 87.8,
        severity: 'Moderate',
        treatments: [
          'Apply chlorothalonil fungicide',
          'Remove infected plant parts',
          'Improve air circulation',
          'Use copper-based sprays',
          'Apply thiophanate-methyl',
        ],
        prevention: [
          'Use disease-free seeds',
          'Avoid overhead watering',
          'Maintain proper spacing',
          'Practice field sanitation',
          'Monitor weather conditions',
        ],
        imageUrl: '/images/anthracnose.jpg',
      },
      {
        diseaseName: 'Root Rot',
        diseaseNameHindi: 'जड़ सड़न',
        description: 'Soil-borne fungal disease causing root decay, wilting, and plant death, especially in waterlogged conditions.',
        confidence: 83.5,
        severity: 'High',
        treatments: [
          'Improve soil drainage',
          'Apply fungicide drench',
          'Remove infected plants',
          'Use biological control agents',
          'Apply carbendazim to soil',
        ],
        prevention: [
          'Ensure proper drainage',
          'Avoid overwatering',
          'Use well-drained soil',
          'Practice crop rotation',
          'Maintain soil pH balance',
        ],
        imageUrl: '/images/root-rot.jpg',
      },
      {
        diseaseName: 'Healthy Plant',
        diseaseNameHindi: 'स्वस्थ पौधा',
        description: 'The plant appears healthy with no visible signs of disease. Continue current care practices.',
        confidence: 94.2,
        severity: 'None',
        treatments: [
          'Continue regular care',
          'Maintain proper nutrition',
          'Monitor regularly',
          'Apply preventive measures',
        ],
        prevention: [
          'Regular field inspection',
          'Proper watering schedule',
          'Balanced fertilization',
          'Good field hygiene',
          'Use quality seeds',
        ],
        imageUrl: '/images/healthy-plant.jpg',
      },
    ];

    // Randomly select a result for demonstration
    const randomIndex = Math.floor(Math.random() * possibleDiseases.length);
    return possibleDiseases[randomIndex];
  }

  // Get all available crops
  getAllCrops(): Crop[] {
    return this.dummyCrops;
  }

  // Search crops by name
  searchCrops(query: string): Crop[] {
    return this.dummyCrops.filter(
      crop =>
        crop.name.toLowerCase().includes(query.toLowerCase()) ||
        crop.nameHindi.includes(query)
    );
  }
}

// Create a singleton instance
export const cropService = new CropService();
